### DISCUSSION TITLE
Computed value

### DISCUSSION THREAD

**Discussion by Jack-Works** [Jack-Works/proposal-enum#12] [General]
Should the value be an expression instead of a literal?

```ts
enum Computed {
    A = expr(),
}
```

## Computed value, but restricted

The computed value is very useful when we want to support the bit-flag enum.

```ts
enum FileAccess {
  None = 0,
  Read = 1 << 1,
  Write = 1 << 2,
  ReadWrite = Read | Write,
}
```

We can limit the ability of computed value to only allows the following on the RHS:

- string, number, bigint literal
- Unary `-` operator (to allow negative number)
- Unary `~` operator (bit-flag)
- Binary `|` `&` `<<` `>>` `>>>`, `^` operator (bit-flag)
- Enum member reference that previously defined in the **current** enum (differ than TypeScript which allows access of different enum).
- `(expr)` to allow specify the priority of computation

## Problem

The full dynamic computed property value conflicts with inline-able (#4) and hoistable declaration (#18).

The limited version conflicts with Heterogeneous Enum (#8) because you cannot declare an enum member should be a symbol. unless we have some ad-hoc syntax for it.

**Comment by ljharb** [comment]
Why would this be restricted? Values must be able to be symbols, first of all, and i consider #8 to be a requirement.

**Reply by Jack-Works** [reply]
This feature conflicts with the statically analyzable requirement (#4). If we give up this requirement, we can have this feature and #8.
